### ISSUE TITLE
ArrayIndexOutOfBoundsException with nested aggregations

### ISSUE THREAD

**Issue by bwebster** [elastic/elasticsearch#16391]
We are running a query that has nested aggregations.  We are experiencing sporadic failures with our test suite around this query.  

We are running 2.1.1.  Here is the version info:

```
{
  number: "2.1.1",
  build_hash: "40e2c53a6b6c2972b3d13846e450e66f4375bd71",
  build_timestamp: "2015-12-15T13:05:55Z",
  build_snapshot: false,
  lucene_version: "5.3.1"
}
```

We are using a time-based index strategy (one index per month), controlled by a template.  The issue seems to be related to time.  If we mock out the current time in our tests so that all data is created in a single monthly index, the issue goes away.  

The query we are running looks like this:

```
{
  "size" : 0,
  "query" : {
    "filtered" : {
      "filter" : {
        "and" : [
          {
            "terms" : {
              "newsroom_id" : [
                "2165011d66f4115cca001815"
              ]
            }
          },
          {
            "range" : {
              "timestamp" : {
                "gte" : "2016-01-03T22:47:03.000Z",
                "lte" : "2016-02-02T22:47:03.000+00:00"
              }
            }
          }
        ]
      }
    }
  },
  "aggs" : {
    "post_id" : {
      "terms" : {
        "field" : "post_id",
        "order" : [
          {
            "opportunities_ae220c23>value.sum" : "desc"
          },
          {
            "all_stats.sum" : "desc"
          }
        ],
        "size" : 0
      },
      "aggs" : {
        "opportunities_ae220c23" : {
          "filter" : {
            "term" : {
              "stage" : "ae220c23"
            }
          },
          "aggs" : {
            "value" : {
              "stats" : {
                "field" : "stats.opportunities"
              }
            }
          }
        },
        "all_stats" : {
          "stats" : {
            "field" : "stats.opportunities"
          }
        }
      }
    }
  }
}
```

In our test suite, we do the following:
- delete all the documents in the index
- create a bunch of new documents
- refresh the indexes
- run the query

This test seems to fail 5 out of 10 times, and we get different failures.  Sometimes we get some data back, but not all the data we are expecting.  Other times we get a `500` response back from ElasticSearch.

I turned up logging and here are a few samples of what I'm seeing:

```
[2016-02-02 15:47:04,003][DEBUG][action.search.type       ] [George Stacy] [analytics_test_201602][0], node[OcKoG1smQ0a__zSHxix33A], [P], v[2], s[STARTED], a[id=_-rGSdgvTtyeXZ1yynsypA]: Failed to execute [org.elasticsearch.action.search.SearchRequest@793d7235]
RemoteTransportException[[George Stacy][127.0.0.1:9300][indices:data/read/search[phase/query]]]; nested: ArrayIndexOutOfBoundsException;
Caused by: java.lang.ArrayIndexOutOfBoundsException
[2016-02-02 15:47:04,003][DEBUG][action.search.type       ] [George Stacy] [analytics_test_201601][0], node[OcKoG1smQ0a__zSHxix33A], [P], v[2], s[STARTED], a[id=ynufly_CRLGMmPEfk0Q1pA]: Failed to execute [org.elasticsearch.action.search.SearchRequest@793d7235] lastShard [true]
RemoteTransportException[[George Stacy][127.0.0.1:9300][indices:data/read/search[phase/query]]]; nested: ArrayIndexOutOfBoundsException;
Caused by: java.lang.ArrayIndexOutOfBoundsException
[2016-02-02 15:47:04,003][DEBUG][action.search.type       ] [George Stacy] All shards failed for phase: [query]
RemoteTransportException[[George Stacy][127.0.0.1:9300][indices:data/read/search[phase/query]]]; nested: ArrayIndexOutOfBoundsException;
Caused by: java.lang.ArrayIndexOutOfBoundsException
[2016-02-02 15:47:04,004][INFO ][rest.suppressed          ] /analytics_test_201601,analytics_test_201602/_search Params: {routing=2165011d66f4115cca001815, index=analytics_test_201601,analytics_test_201602}
Failed to execute phase [query], all shards failed; shardFailures {[OcKoG1smQ0a__zSHxix33A][analytics_test_201601][0]: RemoteTransportException[[George Stacy][127.0.0.1:9300][indices:data/read/search[phase/query]]]; nested: ArrayIndexOutOfBoundsException; }{[OcKoG1smQ0a__zSHxix33A][analytics_test_201602][0]: RemoteTransportException[[George Stacy][127.0.0.1:9300][indices:data/read/search[phase/query]]]; nested: ArrayIndexOutOfBoundsException; }
	at org.elasticsearch.action.search.type.TransportSearchTypeAction$BaseAsyncAction.onFirstPhaseResult(TransportSearchTypeAction.java:228)
	at org.elasticsearch.action.search.type.TransportSearchTypeAction$BaseAsyncAction$1.onFailure(TransportSearchTypeAction.java:174)
	at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:46)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:821)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:799)
	at org.elasticsearch.transport.TransportService$4.onFailure(TransportService.java:361)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: ; nested: ArrayIndexOutOfBoundsException;
	at org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:382)
	at org.elasticsearch.action.search.SearchPhaseExecutionException.guessRootCauses(SearchPhaseExecutionException.java:152)
	at org.elasticsearch.action.search.SearchPhaseExecutionException.getCause(SearchPhaseExecutionException.java:99)
	at java.lang.Throwable.printStackTrace(Throwable.java:665)
	at java.lang.Throwable.printStackTrace(Throwable.java:721)
	at org.apache.log4j.DefaultThrowableRenderer.render(DefaultThrowableRenderer.java:60)
	at org.apache.log4j.spi.ThrowableInformation.getThrowableStrRep(ThrowableInformation.java:87)
	at org.apache.log4j.spi.LoggingEvent.getThrowableStrRep(LoggingEvent.java:413)
	at org.apache.log4j.WriterAppender.subAppend(WriterAppender.java:313)
	at org.apache.log4j.WriterAppender.append(WriterAppender.java:162)
	at org.apache.log4j.AppenderSkeleton.doAppend(AppenderSkeleton.java:251)
	at org.apache.log4j.helpers.AppenderAttachableImpl.appendLoopOnAppenders(AppenderAttachableImpl.java:66)
	at org.apache.log4j.Category.callAppenders(Category.java:206)
	at org.apache.log4j.Category.forcedLog(Category.java:391)
	at org.apache.log4j.Category.log(Category.java:856)
	at org.elasticsearch.common.logging.log4j.Log4jESLogger.internalInfo(Log4jESLogger.java:125)
	at org.elasticsearch.common.logging.support.AbstractESLogger.info(AbstractESLogger.java:90)
	at org.elasticsearch.rest.BytesRestResponse.convert(BytesRestResponse.java:131)
	at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:96)
	at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:87)
	at org.elasticsearch.rest.action.support.RestActionListener.onFailure(RestActionListener.java:60)
	at org.elasticsearch.action.search.type.TransportSearchTypeAction$BaseAsyncAction.raiseEarlyFailure(TransportSearchTypeAction.java:316)
	... 10 more
Caused by: java.lang.ArrayIndexOutOfBoundsException
```

```
RemoteTransportException[[George Stacy][127.0.0.1:9300][indices:data/read/search[phase/query]]]; nested: ArrayIndexOutOfBoundsException[1];
Caused by: java.lang.ArrayIndexOutOfBoundsException: 1
	at org.elasticsearch.common.util.BigArrays$DoubleArrayWrapper.get(BigArrays.java:260)
	at org.elasticsearch.search.aggregations.metrics.stats.StatsAggegator.metric(StatsAggegator.java:135)
	at org.elasticsearch.search.aggregations.bucket.terms.InternalOrder$Aggregation$2.compare(InternalOrder.java:200)
	at org.elasticsearch.search.aggregations.bucket.terms.InternalOrder$Aggregation$2.compare(InternalOrder.java:197)
	at org.elasticsearch.search.aggregations.bucket.terms.InternalOrder$CompoundOrder$CompoundOrderComparator.compare(InternalOrder.java:280)
	at org.elasticsearch.search.aggregations.bucket.terms.InternalOrder$CompoundOrder$CompoundOrderComparator.compare(InternalOrder.java:266)
	at org.elasticsearch.search.aggregations.bucket.terms.support.BucketPriorityQueue.lessThan(BucketPriorityQueue.java:37)
	at org.elasticsearch.search.aggregations.bucket.terms.support.BucketPriorityQueue.lessThan(BucketPriorityQueue.java:26)
	at org.apache.lucene.util.PriorityQueue.upHeap(PriorityQueue.java:258)
	at org.apache.lucene.util.PriorityQueue.add(PriorityQueue.java:135)
	at org.apache.lucene.util.PriorityQueue.insertWithOverflow(PriorityQueue.java:151)
	at org.elasticsearch.search.aggregations.bucket.terms.GlobalOrdinalsStringTermsAggregator.buildAggregation(GlobalOrdinalsStringTermsAggregator.java:176)
	at org.elasticsearch.search.aggregations.AggregationPhase.execute(AggregationPhase.java:142)
	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:112)
	at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:363)
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:375)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryTransportHandler.messageReceived(SearchServiceTransportAction.java:368)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryTransportHandler.messageReceived(SearchServiceTransportAction.java:365)
	at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:350)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

I'm also seeing this in the logs:

```
[2016-02-02 16:05:12,166][DEBUG][cluster.action.shard     ] [George Stacy] [analytics_test_201602][4] sending shard started for [analytics_test_201602][4], node[OcKoG1smQ0a__zSHxix33A], [P], v[1], s[INITIALIZING], a[id=ZgJoAG_vRGO6pVamLpHOHg], unassigned_info[[reason=INDEX_CREATED], at[2016-02-02T23:05:12.087Z]], expected_shard_size[7404], indexUUID [_WBlFe96TRO0kbG7S3ZgDg], message [after recovery from store], failure [Unknown]
```

Is anybody experiencing similar issues?

For reference, the data we create for this test looks like this:

```
{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 10,
    "successful": 10,
    "failed": 0
  },
  "hits": {
    "total": 5,
    "max_score": 1,
    "hits": [
      {
        "_index": "analytics_test_201601",
        "_type": "content_scoring_stats",
        "_id": "8e1db69dd42f8eece90b97f129537cdca7eabfd0e4110d9a5cfa9c3703bf0708",
        "_score": 1,
        "_routing": "2165011d66f4115cca001815",
        "_source": {
          "created_at": "2016-02-02T22:47:06.990Z",
          "updated_at": "2016-02-02T22:47:06.990Z",
          "job_id": "2bc0c0dd-5a38-42d5-8961-22510ae6fdeb",
          "request_id": null,
          "timestamp": "2016-01-31T22:47:06.990Z",
          "recorded_at": "2016-02-02T22:47:00.548+00:00",
          "newsroom_id": "2165011d66f4115cca001815",
          "post_id": "d0c672a0e5f0f66e906ccd90",
          "stage": "11494a78",
          "stats": {
            "opportunities": 5.5,
            "revenue": 100.5
          }
        }
      },
      {
        "_index": "analytics_test_201601",
        "_type": "content_scoring_stats",
        "_id": "0e6e49d68abe67150cee093c38fe2acdcec88a6491b862a3a80adcf348a4df1b",
        "_score": 1,
        "_routing": "2165011d66f4115cca001815",
        "_source": {
          "created_at": "2016-02-02T22:47:06.992Z",
          "updated_at": "2016-02-02T22:47:06.992Z",
          "job_id": "3a19c4e6-8446-43c4-a6b6-ed471411129f",
          "request_id": null,
          "timestamp": "2016-01-31T22:47:06.992Z",
          "recorded_at": "2016-02-02T22:47:00.548+00:00",
          "newsroom_id": "2165011d66f4115cca001815",
          "post_id": "bd2b0c5993413414f1f92f3d",
          "stage": "f89d57cb",
          "stats": {
            "opportunities": 10.2,
            "revenue": 24.3
          }
        }
      },
      {
        "_index": "analytics_test_201602",
        "_type": "content_scoring_stats",
        "_id": "2982db4c6f50ca5611f331f51474a882d32ea40b628db51eca1f8beed8484e35",
        "_score": 1,
        "_routing": "2165011d66f4115cca001815",
        "_source": {
          "created_at": "2016-02-02T22:47:06.995Z",
          "updated_at": "2016-02-02T22:47:06.995Z",
          "job_id": "406b92b1-2b05-4cb1-93af-7651021d25d5",
          "request_id": null,
          "timestamp": "2016-02-01T22:47:06.995Z",
          "recorded_at": "2016-02-02T22:47:00.548+00:00",
          "newsroom_id": "2165011d66f4115cca001815",
          "post_id": "d0c672a0e5f0f66e906ccd90",
          "stage": "11494a78",
          "stats": {
            "opportunities": 0.3,
            "revenue": 43.1
          }
        }
      },
      {
        "_index": "analytics_test_201602",
        "_type": "content_scoring_stats",
        "_id": "c7140c8a6ebb4422b5ab1c495c6afca517ccbd5ea284e3aa0d6755320cba0426",
        "_score": 1,
        "_routing": "2165011d66f4115cca001815",
        "_source": {
          "created_at": "2016-02-02T22:47:06.997Z",
          "updated_at": "2016-02-02T22:47:06.997Z",
          "job_id": "3ca37f8c-b143-46f7-916b-5fbd2f06c653",
          "request_id": null,
          "timestamp": "2016-02-01T22:47:06.997Z",
          "recorded_at": "2016-02-02T22:47:00.548+00:00",
          "newsroom_id": "2165011d66f4115cca001815",
          "post_id": "c766f88321f126c2309ce791",
          "stage": "f89d57cb",
          "stats": {
            "opportunities": 0.5,
            "revenue": 1.8
          }
        }
      },
      {
        "_index": "analytics_test_201602",
        "_type": "content_scoring_stats",
        "_id": "4b1cd5c7df3aa55879253021ef59ac73975736652763eef8f798ea50ac004ef5",
        "_score": 1,
        "_routing": "e2d34791522c86d9210673f2",
        "_source": {
          "created_at": "2016-02-02T22:47:06.999Z",
          "updated_at": "2016-02-02T22:47:06.999Z",
          "job_id": "56e1a19d-b8be-42ba-8912-aff1084bf347",
          "request_id": null,
          "timestamp": "2016-02-01T22:47:06.999Z",
          "recorded_at": "2016-02-02T22:47:00.548+00:00",
          "newsroom_id": "e2d34791522c86d9210673f2",
          "post_id": "c766f88321f126c2309ce791",
          "stage": "f89d57cb",
          "stats": {
            "opportunities": 10,
            "revenue": 30
          }
        }
      }
    ]
  }
}
```

**Comment by jasontedor** [comment]
Someone else will look at the core of the issue here but I just wanted to let you know that the message

> I'm also seeing this in the logs:
> ```
[2016-02-02 16:05:12,166][DEBUG][cluster.action.shard     ] [George Stacy] [analytics_test_201602][4] sending shard started for [analytics_test_201602][4], node[OcKoG1smQ0a__zSHxix33A], [P], v[1], s[INITIALIZING], a[id=ZgJoAG_vRGO6pVamLpHOHg], unassigned_info[[reason=INDEX_CREATED], at[2016-02-02T23:05:12.087Z]], expected_shard_size[7404], indexUUID [_WBlFe96TRO0kbG7S3ZgDg], message [after recovery from store], failure [Unknown]
```

that you pointed out is fine, it's not a failure, it's just bad logging. In this case, a shard is being started after recovering. However, shard started and shard failure actions share the same request object and their shared string representation prints out "failure: [Unknown]" when there is no failure. It's bothered me for awhile, but seeing it here finally brought me to do something about it; I opened #16396. Sorry for the confusion!

**Comment by bwebster** [comment]
Thanks for the explanation @jasontedor.  Good to know that the `failure: [Unknown]` isn't something to worry about.

**Comment by bwebster** [comment]
Update: the test suite is now much more reliable.  The only difference is that the test data is no longer overlapping a month boundary.  All the test data is created in one index (for the month of February).

Seems to be related to the query needing to hit more than one index.

**Comment by 2e3s** [comment]
I have the same error, narrowed it if it helps, the different parts are, query with the error:
```json
  "aggregations": {
    "root": {
      "terms": {
        "field": "result",
        "order": [
          {
            "current>amount": "desc"
          }
        ]
      },
```
with no error
```json
  "aggregations": {
    "root": {
      "terms": {
        "field": "result"
      },
```

**Comment by colings86** [comment]
This looks to be the same as https://github.com/elastic/elasticsearch/issues/17225 which has been fixed by https://github.com/elastic/elasticsearch/pull/17379 and should be available from version 2.3.1. Could you confirm whether you still see this on the latest 2.3 branch?

**Comment by clintongormley** [comment]
No further feedback. Closing
